sdk/servicebus/service-bus/src/clientEntityContext.ts

```diff
@@ -200,26 +219,76 @@ export namespace ClientEntityContext {
       }
     };
 
-    (entityContext as ClientEntityContext).clearClientReference = async (clientId: string) => {
-      delete context.clients[clientId];
+    const isManagementClientSharedWithOtherClients = (): boolean => {
+      for (const id of Object.keys(context.clientContexts)) {
+        if (
+          context.clientContexts[id].entityPath === entityContext.entityPath &&
+          context.clientContexts[id].clientId !== entityContext.clientId
+        ) {
+          return true;
+        }
+      }
+      return false;
+    };
 
-      if (!entityContext.managementClient) {
+    (entityContext as ClientEntityContext).close = async () => {
+      if (!context.connection || !context.connection.isOpen()) {
         return;
       }
-      let isManagementClientInUse = false;
-      for (const id of Object.keys(context.clients)) {
-        if (context.clients[id].entityPath === entityContext.entityPath) {
-          isManagementClientInUse = true;
-          break;
-        }
+
+      log.entityCtxt(
+        "[%s] Closing client entity context for %s: %O",
+        context.connectionId,
+        clientId
+      );
+
+      // Close sender
+      if (entityContext.sender) {
+        await entityContext.sender.close();
+      }
+
+      // Close batching receiver
+      if (entityContext.batchingReceiver) {
+        await entityContext.batchingReceiver.close();
+      }
+
+      // Close streaming receiver
+      if (entityContext.streamingReceiver) {
+        await entityContext.streamingReceiver.close();
+      }
+
+      // Close all the MessageSessions.
+      for (const messageSessionId of Object.keys(entityContext.messageSessions)) {
+        await entityContext.messageSessions[messageSessionId].close();
+      }
+
+      // Close the sessionManager.
+      if (entityContext.sessionManager) {
+        entityContext.sessionManager.close();
       }
-      if (!isManagementClientInUse) {
+
+      // Make sure that we clear the map of deferred messages
+      entityContext.requestResponseLockedMessages.clear();
+
+      // Delete the reference in ConnectionContext
+      delete context.clientContexts[clientId];
+
+      // Close the managementClient unless it is shared with other clients
+      if (entityContext.managementClient && isManagementClientSharedWithOtherClients()) {
```

Comment:
Comment says unless, but code reads if to me. Did you forget a bang?

Reply:
Yes, I did!
And the tests are all passing, but the test won't exit because of this.
Was trying to figure out the cause for the non exiting tests and found exactly this.
Pushed a commit to fix it.
Should have seen your comment earlier :(